### ISSUE TITLE
Asset hot reloading

### ISSUE THREAD

**Issue by torkleyy** [amethyst/amethyst#266]
I've figured most parts out already, however I'm stuck at the last one. After a file has changed, it's been read, parsed and turned into an asset, how should we replace the old assets?

Most likely assets live as components in the world, but still, how should we track them? Or is limiting it to the `World` a good idea at all?

I'm not sure whether or not we need this feature before 0.5 release, probably that needs discussion, too.

**Comment by Xaeroxe** [comment]
In order to prevent memory corruption issues we'll have to pause execution while the replacement occurs.  My first reaction is to put assets in an Arc<RwLock<T>> so we can track them and still hand out references to them.  We could keep some kind of Boolean "dirty" bit representing that assets are ready to be replaced, and check that every frame.  If they are ready to be replaced then pop the new assets from a queue and replace each one until the queue is empty.

**Comment by omni-viral** [comment]
Global pauses are unnecessary. There data structures which permit update of value while it is read.
We should use one that works for one writer and many readers.

**Comment by Xaeroxe** [comment]
@omni-viral I don't know of any such data structures.  Can you provide one for me to read about?

**Comment by omni-viral** [comment]
@Xaeroxe I'll find link  to the paper to post it here.
Our case has quiescent points. So we can search in that direction.

**Comment by omni-viral** [comment]
Maybe there is no need of fancy data structures either if we can update assets between systems run

**Comment by omni-viral** [comment]
Maybe there is no need in fancy data structures if we can update assets between systems run

**Comment by torkleyy** [comment]
Still, how would we trace the assets?

**Comment by Xaeroxe** [comment]
We would need some kind of central `AssetManager` structure that keeps a `HashMap<String, Arc<RwLock<T>>>` so it can be referred to when the time to replace the asset comes.

**Comment by Xaeroxe** [comment]
(Said structure would also require a way to remove references so that the memory the assets consume can be freed)

**Comment by omni-viral** [comment]
Put all updates in a queue. Replace the queue atomically in maintenance phase. Flush updates into the world.

**Comment by omni-viral** [comment]
@Xaeroxe no locks please)

**Comment by Xaeroxe** [comment]
Then the update will have to occur in unsafe code, which is fine but we'll have to be much more careful about memory corruption.

**Comment by omni-viral** [comment]
Well. We should wrap unsafe code into nice and safe type and test it heavily

**Comment by torkleyy** [comment]
Not sure if the result is something we want, but what you're asking for
sounds like shred's `TrustCell`.

**Comment by Xaeroxe** [comment]
Ok here's a nice approach that gives us minimal pausing: Arc<Cell<Box<T>>>

Rather than locking the value while it is replaced you just atomically call "set" on the cell and replace the Box value.

**Comment by Xaeroxe** [comment]
Except that doesn't work because Box doesn't implement Copy.  Yeah we'll have to use unsafe code.

**Comment by torkleyy** [comment]
The std `Cell` only works for `Copy` types IIRC.

**Comment by omni-viral** [comment]
@torkleyy can't find any docs on `TrustCell`

**Comment by Xaeroxe** [comment]
@omni-viral It's not part of the official API, look [here](https://github.com/slide-rs/shred/blob/master/src/cell.rs#L72)

**Comment by omni-viral** [comment]
@torkleyy `TrustCell` would work find for asset holder inside `World` for async read acces

**Comment by omni-viral** [comment]
One more question. How many asset producers will be there? Single or multiple?

**Comment by Xaeroxe** [comment]
@omni-viral Current code permits multiple, which is nice for asynchronous loading from multiple systems but we may require some kind of unification to implement hot reloading.

**Comment by omni-viral** [comment]
I'm suggesting following algorithm.
Have `Arc<Mutex<SomeNiceQueue<AssetUpdate>>>` inside `Engine`
Store all updates in this queue. Blocking updating threads is OK I think.
At maintenance phase (between systems run) swap queue with empty one. And flush all updates into `TrustCell`s

**Comment by omni-viral** [comment]
`Mutex` in `Arc<Mutex<Vec<AssetUpdate>>>` could be replaced with something that will allow not to block main thread.

**Comment by Xaeroxe** [comment]
I disagree, that's not performance critical code and in my opinion it doesn't warrant the risk of unsafety to replace Mutex in that type with a non-locking unsafe variant.

**Comment by Xaeroxe** [comment]
Alternatively if blocking is really that undesirable, we could use a [channel](https://doc.rust-lang.org/std/sync/mpsc/) to send the `AssetUpdate`s

**Comment by omni-viral** [comment]
@Xaeroxe mpsc channel is the data structure with internal unsafe you were opposed to one comment earlier 😄

**Comment by Xaeroxe** [comment]
Yeah I just thought you meant some kind of TrustCell for the queue, which seemed excessive to me.  Channel works much better though :)

**Comment by torkleyy** [comment]
The thing is, I'm not sure how deeply hot reloading should be integrated in
the design, since it is a development feature, not something that's
necessary for the actual game.

**Comment by Xaeroxe** [comment]
That is a good point.  In order to make the feature work we need some kind of tracking and continuous updating though.  This seems to be the easiest way to accomplish that.  Perhaps we could use `#[cfg(build = "debug")]` to prevent hot reloading code from being in the final game.

**Comment by omni-viral** [comment]
`#[cfg(build = "debug")]` isn't best here. You probably want to run your game with release build during development cause it may run too slow in debug.
Instead there should be distinct feature.

**Comment by torkleyy** [comment]
That's what we planned to use `amethyst_release` for.

**Comment by Rhuagh** [comment]
If hot reloading is a pure development feature, some risk might be worth it, for a simpler solution?

**Comment by Xaeroxe** [comment]
Ok, what are you proposing?  Mpsc channel seems to be a best of all worlds approach.  It's a non-locking std library type with internal unsafe and no need for us to write unsafe sections of code.

**Comment by Rhuagh** [comment]
Oh, I don't know what would be best, my suggestion is just to do the simplest solution that fits the bill if it's only gonna be used for dev, and not for released product :)

**Comment by torkleyy** [comment]
So essentially that problem we discussed about (updating the asset) disappeared. This was part of my second assets rework, which introduced asset storages and handles, allowing us to control reads and writes in a central place.

**Comment by erlend-sh** [comment]
Fair amount of discussion about this in the Rust community recently thanks to @phaazon's [warmy](https://github.com/phaazon/warmy).

Seems @icefoxen is already using it successfully in ggez.

**Comment by torkleyy** [comment]
@erlend-sh I mostly like what we have currently, it fits nicely into ECS and works in parallel. `warmy` has dependencies, which is something that's still missing.

But at a first glance we couldn't even use `warmy` because it's not thread-safe.

**Comment by phaazon** [comment]
`warmy` is not thread-safe **yet**. :)
